reactive/canal.py

```diff
@@ -29,16 +29,19 @@ def upgrade_charm():
         hookenv.log(e)
 
 
-@when('etcd.available', 'cni.connected', 'flannel.service.started',
+@when('etcd.available', 'cni.configured', 'flannel.service.started',
       'calico.service.started', 'calico.pool.configured')
 @when_not('canal.cni.configured')
 def configure_cni():
     ''' Configure Calico CNI. '''
     status_set('maintenance', 'Configuring Calico CNI')
     os.makedirs('/etc/cni/net.d', exist_ok=True)
-    cni = endpoint_from_flag('cni.connected')
+    cni = endpoint_from_flag('cni.configured')
     etcd = endpoint_from_flag('etcd.available')
     cni_config = cni.get_config()
+    if 'kubeconfig_path' not in cni_config:
+        hookenv.log('Unable to get configuration for cni. Will retry.')
+        return
```

Comment:
This check prevents canal from coming up when related to kubernetes-master, since kubernetes-master only sets kubeconfig_path to `''`: https://github.com/juju-solutions/kubernetes/blob/55e305871069929399d627cdfd51479ad34fa5a0/cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py#L318

The `kubeconfig_path` key doesn't end up in the relation data at all:
```
$ juju run --unit canal/0 -- relation-get -r cni:11 - kubernetes-master/1
egress-subnets: 172.31.8.252/32
ingress-address: 172.31.8.252
is_master: "True"
private-address: 172.31.8.252
```

I think you can safely remove this check. Waiting on the new `cni.configured` flag should be enough to address the original issue right? Although, removing this check would mean the master will go ahead and generate a cni .conf that we know is bad. Not a big deal since the master doesn't use that .conf anyway, but it does seem kinda weird, huh?

Reply:
Yeah, I was torn on this and at one point had a change for the master to add a full kubeconfig. That was my main test subject for this. My (incorrect) assumption was that it would show up as an empty string for the master. I will clean my cache and make sure that this works.